### ISSUE TITLE
[BUG]  Error: Did not receive X-CSRF-TOKEN

### ISSUE THREAD

**Issue by JonasTheGamer** [suufi/noblox.js#249]
## Before posting a bug, be sure that someone else has not posted it already and it is an actual bug with noblox.js and not other frameworks like discord.js

**Describe the bug**
Hi, I'm trying to make my own addition to the node module (login.js). The idea is that I submit the username, password and captchaToken. That has worked before (Around a year ago). However when trying to redo this (I accidentally deleted my old code) and updating the module it looks like I'm not getting an X-CSRF-Token.  This is the full error:
`(node:17400) UnhandledPromiseRejectionWarning: Error: Did not receive X-CSRF-TOKEN
    at C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\noblox.js\lib\util\getGeneralToken.js:30:15
    at tryCatcher (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\util.js:16:23)
    at Promise._settlePromiseFromHandler (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\promise.js:547:31)
    at Promise._settlePromise (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\promise.js:604:18)
    at Promise._settlePromise0 (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\promise.js:729:18)
    at _drainQueueStep (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\async.js:93:12)
    at _drainQueue (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\async.js:86:9)
    at Async._drainQueues (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (C:\Users\jonas\OneDrive\Bureaublad\Folders\bloxtech nodejs logindetails to cookie\node_modules\bluebird\js\release\async.js:15:14)
    at processImmediate (internal/timers.js:439:21)
(node:17400) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:17400) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

**My login.js code (I took some from the old login script :) )**
`
// Includes
var http = require('../util/http.js').func
var options = require('../options.js')
var settings = require('../../settings.json')
var clearSession = require('../util/clearSession.js').func
var getGeneralToken = require('../util/getGeneralToken.js').func
var getJar = require('../util/jar.js').func

// Args
exports.required = ['username', 'password','extratext']
exports.optional = ['jar']

// Define
function login (jar, token, username, password, extratext) {
  console.log(jar);
  console.log(token);
  console.log(username);
  console.log(password);

  clearSession({ jar: jar })
  var url = '//auth.roblox.com/v2/login'
  var post = {
    ctype: 'Username',
    cvalue: username,
    password: password,
    captchaToken: extratext,
    captchaProvider: 'PROVIDER_ARKOSE_LABS'
  }
  var httpOpt = {
    url: url,
    options: {
      method: 'POST',
      json: post,
      resolveWithFullResponse: true,
      jar: jar,
      headers: {
        'X-CSRF-TOKEN': token
      }
    }
  }
  return http(httpOpt)
    .then(function (res) {
      var body = res.body
      if (res.statusCode === 200) {
        if (body.twoStepVerificationData) {
          throw new Error('Two step verification is not supported')
        }
        if (settings.session_only) {
          var cookies = res.headers['set-cookie'] // If the user is already logged in a new cookie will not be returned.
          if (cookies) {
            var session = cookies.toString().match(/\.ROBLOSECURITY=(.*?);/)[1]
            jar.session = session
          }
        }
        return body.user
      } else {
        if (body.errors && body.errors.length > 0) {
          var errors = body.errors.map((e) => {
            return e.message
          })
          throw new Error('Login failed: ' + errors.join(', '))
        }
      }
    })
}

exports.func = function (args) {
  var jar = args.jar || options.jar
  var username = args.username
  var password = args.password
  var extratext = args.extratext
  return getGeneralToken({ jar: getJar() }) // Get token is as if you are not logged in
    .then(function (xcsrf) {
      return login(jar, xcsrf, username, password, extratext)
    })
}
`

**Expected behavior**
I'd expect this code to login like it should but instead it returns me that error. (I checked the getGeneralToken script, that's the same one as I can see here on Github now.)
**Screenshots**
![image](https://user-images.githubusercontent.com/38383712/103530180-46c81000-4e87-11eb-9ea2-1e68326e3add.png)

Thanks in advance for your help! :-)

**Comment by ChrisvanChip** [comment]
Formatted JavaScript:
`// Includes
var http = require('../util/http.js').func
var options = require('../options.js')
var settings = require('../../settings.json')
var clearSession = require('../util/clearSession.js').func
var getGeneralToken = require('../util/getGeneralToken.js').func
var getJar = require('../util/jar.js').func

// Args
exports.required = ['username', 'password', 'extratext']
exports.optional = ['jar']

// Define
function login(jar, token, username, password, extratext) {
    console.log(jar);
    console.log(token);
    console.log(username);
    console.log(password);

    clearSession({
        jar: jar
    })
    var url = '//auth.roblox.com/v2/login'
    var post = {
        ctype: 'Username',
        cvalue: username,
        password: password,
        captchaToken: extratext,
        captchaProvider: ''
    }
    var httpOpt = {
        url: url,
        options: {
            method: 'POST',
            json: post,
            resolveWithFullResponse: true,
            jar: jar,
            headers: {
                'X-CSRF-TOKEN': token
            }
        }
    }
    return http(httpOpt)
        .then(function(res) {
            var body = res.body
            if (res.statusCode === 200) {
                if (body.twoStepVerificationData) {
                    throw new Error('Two step verification is not supported')
                }
                if (settings.session_only) {
                    var cookies = res.headers['set-cookie'] // If the user is already logged in a new cookie will not be returned.
                    if (cookies) {
                        var session = cookies.toString().match(/.ROBLOSECURITY=(.*?);/)[1]
                        jar.session = session
                    }
                }
                return body.user
            } else {
                if (body.errors && body.errors.length > 0) {
                    var errors = body.errors.map((e) => {
                        return e.message
                    })
                    throw new Error('Login failed: ' + errors.join(', '))
                }
            }
        })
}

exports.func = function(args) {
    var jar = args.jar || options.jar
    var username = args.username
    var password = args.password
    var extratext = args.extratext
    return getGeneralToken({
            jar: getJar()
        }) // Get token is as if you are not logged in
        .then(function(xcsrf) {
            return login(jar, xcsrf, username, password, extratext)
        })
}`

**Comment by ChrisvanChip** [comment]
Formatted JavaScript:
```// Includes
var http = require('../util/http.js').func
var options = require('../options.js')
var settings = require('../../settings.json')
var clearSession = require('../util/clearSession.js').func
var getGeneralToken = require('../util/getGeneralToken.js').func
var getJar = require('../util/jar.js').func

// Args
exports.required = ['username', 'password', 'extratext']
exports.optional = ['jar']

// Define
function login(jar, token, username, password, extratext) {
    console.log(jar);
    console.log(token);
    console.log(username);
    console.log(password);

    clearSession({
        jar: jar
    })
    var url = '//auth.roblox.com/v2/login'
    var post = {
        ctype: 'Username',
        cvalue: username,
        password: password,
        captchaToken: extratext,
        captchaProvider: ''
    }
    var httpOpt = {
        url: url,
        options: {
            method: 'POST',
            json: post,
            resolveWithFullResponse: true,
            jar: jar,
            headers: {
                'X-CSRF-TOKEN': token
            }
        }
    }
    return http(httpOpt)
        .then(function(res) {
            var body = res.body
            if (res.statusCode === 200) {
                if (body.twoStepVerificationData) {
                    throw new Error('Two step verification is not supported')
                }
                if (settings.session_only) {
                    var cookies = res.headers['set-cookie'] // If the user is already logged in a new cookie will not be returned.
                    if (cookies) {
                        var session = cookies.toString().match(/.ROBLOSECURITY=(.*?);/)[1]
                        jar.session = session
                    }
                }
                return body.user
            } else {
                if (body.errors && body.errors.length > 0) {
                    var errors = body.errors.map((e) => {
                        return e.message
                    })
                    throw new Error('Login failed: ' + errors.join(', '))
                }
            }
        })
}

exports.func = function(args) {
    var jar = args.jar || options.jar
    var username = args.username
    var password = args.password
    var extratext = args.extratext
    return getGeneralToken({
            jar: getJar()
        }) // Get token is as if you are not logged in
        .then(function(xcsrf) {
            return login(jar, xcsrf, username, password, extratext)
        })
}```

**Comment by ChrisvanChip** [comment]
Potentially related to #163, #165 and #196.

**Comment by JonasTheGamer** [comment]
Useful info: I'm using v4.7.3

**Comment by Neztore** [comment]
This is not really a Noblox issue given you are setting the session cookie yourself.
Are you sure the cookie is being set into the jar session? Are you sure it is being extracted correctly?
What if you try running setCookie with the cookie you fetch?

**Comment by ChrisvanChip** [comment]
We have moved the code out of `index.js` to the `server.js` file and manually used the `http` function from the library. We made use of `setCookie` to manually set a valid cookie, as it turns out [a Roblox cookie has to be set](https://devforum.roblox.com/t/roblox-new-way-of-getting-x-csrf-token/871979/18?u=xchris_vc) in order to get a CSRF token.

**Comment by Neztore** [comment]
That is correct. Roblox will not return it without it - This is why the getGeneralToken function throws an error if valid a cookie is not set.

**Comment by JonasTheGamer** [comment]
Thanks for the quick assistance! Sad to see it's not possible without an existing cookie but we'll go with it haha. :-) Closing this issue.

TLDR: You'd need an account to be logged in already via `setCookie` which will then allow you to login using username, password and captchaToken.
